File: src/pci/virtio/balloon/balloon.c
#include "balloon.h"
#include "mm/mmu/mmu.h"
#include "mm/pmm/pmm.h"
#include "mmio/mmio.h"
#include "uart/uart.h"
#include "utils/utils.h"

/* ---------------------------------------------------------------------------
 * virtio-balloon — cooperative memory ballooning
 *
 * The host can ask the guest to "inflate" the balloon, meaning the guest
 * should hand a number of physical pages back to the host (the host can
 * then reuse that physical memory elsewhere). The host can later ask the
 * guest to "deflate", meaning the guest may reclaim those pages.
 *
 * Spec mechanics we care about:
 *   - device_cfg.num_pages — host writes the *target* balloon size here
 *   - device_cfg.actual    — driver writes the *current* balloon size
 *   - inflateq (vq 0)      — driver submits arrays of PFNs it just gave away
 *   - deflateq (vq 1)      — driver submits PFNs it is reclaiming
 *   - PFN is a flat phys_addr >> 12, regardless of guest/host page size
 *
 * Driver policy in this kernel:
 *   The kernel does not (yet) wire a virtio-config-change interrupt, so
 *   inflate/deflate are *driver-initiated*: the shell exposes
 *   `balloon inflate N` / `balloon deflate N`, and /proc/balloon shows
 *   both the actual size and the host's wish so the user can spot drift.
 * --------------------------------------------------------------------------- */

/* Page-aligned virtqueue backing memory — same pattern as rng/blk. We
 * have two queues, hence two of each ring. */
static struct virtq_desc inflate_desc[VIRTQ_MAX_SIZE]
    __attribute__((aligned(4096)));
static struct virtq_avail inflate_avail __attribute__((aligned(4096)));
static struct virtq_used inflate_used __attribute__((aligned(4096)));

static struct virtq_desc deflate_desc[VIRTQ_MAX_SIZE]
    __attribute__((aligned(4096)));
static struct virtq_avail deflate_avail __attribute__((aligned(4096)));
static struct virtq_used deflate_used __attribute__((aligned(4096)));

/* DMA-visible buffer the device reads PFNs out of. One u32 per PFN, capped
 * at MAX_PAGES, so 4 KiB max — fits in exactly one page. */
static uint32_t pfn_buf[VIRTIO_BALLOON_MAX_PAGES] __attribute__((aligned(64)));

/* Backing book-keeping: which pages we currently have given to the host.
 * Stored as PFNs to match the wire format. The active prefix is
 * pfns[0 .. actual-1]; everything beyond is undefined. */
static uint32_t inflated_pfns[VIRTIO_BALLOON_MAX_PAGES];

static struct virtio_balloon bln_dev;
static int bln_ready = 0;

/* ---------------------------------------------------------------------------
 * Helpers
 * --------------------------------------------------------------------------- */

/* Push a PFN array on a queue and wait for the device to ack. The buffer
 * is *device-readable* (no VIRTQ_DESC_F_WRITE) — the driver is informing
 * the device which physical pages belong to / are being reclaimed from
 * the balloon. */
static int submit_pfn_batch(struct virtqueue *vq, uint32_t count) {
  if (count == 0) {
    return 0;
  }
  uint64_t pa = VIRT_TO_PHYS((uint64_t)(uintptr_t)pfn_buf);
  virtqueue_submit(vq, pa, count * sizeof(uint32_t), VIRTQ_DESC_F_NONE);
  virtqueue_notify(vq);
  /* virtqueue_poll returns 0 on timeout. The device writes nothing into
   * a read-only buffer so a successful ack also reports 0 bytes — we
   * cannot use the return value to distinguish success from timeout.
   * However virtqueue_poll already logs and returns on timeout, and on
   * success it advances last_used. The simplest reliable signal is to
   * just call it and trust the spin counter inside. */
  (void)virtqueue_poll(vq);
  return 0;
}

/* Mirror our internal balloon size into device_cfg.actual so the host's
 * stats reflect reality. Spec §5.5.6: this is informational. */
static void publish_actual(void) {
  uintptr_t cfg = bln_dev.pci_caps.device_cfg;
  if (cfg == 0) {
    return; /* device_cfg cap not advertised — shouldn't happen for balloon */
  }
  mmio_write32(cfg + VIRTIO_BALLOON_CFG_ACTUAL, bln_dev.actual);
  dsb_sy();
}

/* ---------------------------------------------------------------------------
 * Public API
 * --------------------------------------------------------------------------- */

int balloon_inflate(uint32_t n) {
  if (!bln_ready) {
    return -1;
  }
  /* Cap to the headroom in our tracking array. */
  uint32_t headroom = VIRTIO_BALLOON_MAX_PAGES - bln_dev.actual;
  if (n > headroom) {
    n = headroom;
  }
  if (n == 0) {
    return 0;
  }

  /* Allocate fresh pages from the PMM and stage their PFNs into pfn_buf
   * for the device, while also recording them into our inflated_pfns
   * tracking table so we can hand them back later. */
  uint32_t got = 0;
  for (uint32_t i = 0; i < n; i++) {
    uintptr_t pa = pmm_allocate_page();
    if (!pa) {
      break; /* OOM — stop here, ship what we have */
    }
    uint32_t pfn = (uint32_t)(pa >> VIRTIO_BALLOON_PFN_SHIFT);
    pfn_buf[got] = pfn;
    inflated_pfns[bln_dev.actual + got] = pfn;
    got++;
  }
  if (got == 0) {
    return 0; /* PMM was completely empty */
  }

  if (submit_pfn_batch(&bln_dev.inflate_vq, got) < 0) {
    /* Device handshake failed — give the pages back to the PMM and bail.
     * Our tracking table was speculative so we just ignore those slots. */
    for (uint32_t i = 0; i < got; i++) {
      pmm_free_page((uintptr_t)pfn_buf[i] << VIRTIO_BALLOON_PFN_SHIFT);
    }
    return -1;
  }

  bln_dev.actual += got;
  publish_actual();
  return (int)got;
}

int balloon_deflate(uint32_t n) {
  if (!bln_ready) {
    return -1;
  }
  if (n > bln_dev.actual) {
    n = bln_dev.actual;
  }
  if (n == 0) {
    return 0;
  }

  /* Pop the *most recently* inflated PFNs (LIFO). Stage them into pfn_buf
   * so the device sees the same handles it received on inflateq. */
  uint32_t base = bln_dev.actual - n;
  for (uint32_t i = 0; i < n; i++) {
    pfn_buf[i] = inflated_pfns[base + i];
  }

  if (submit_pfn_batch(&bln_dev.deflate_vq, n) < 0) {
    return -1; /* leave the balloon as-is */
  }

  /* Hand the physical pages back to the PMM. The host has acknowledged
   * that it's no longer using them. */
  for (uint32_t i = 0; i < n; i++) {
    pmm_free_page((uintptr_t)pfn_buf[i] << VIRTIO_BALLOON_PFN_SHIFT);
  }

  bln_dev.actual -= n;
  publish_actual();
  return (int)n;
}

void balloon_get_status(uint32_t *actual_pages, uint32_t *host_target) {
  if (actual_pages) {
    *actual_pages = bln_ready ? bln_dev.actual : 0;
  }
  if (host_target) {
    if (!bln_ready || bln_dev.pci_caps.device_cfg == 0) {
      *host_target = 0;
    } else {
      *host_target = mmio_read32(bln_dev.pci_caps.device_cfg +
                                 VIRTIO_BALLOON_CFG_NUM_PAGES);
    }
  }
}

/* ---------------------------------------------------------------------------
 * Init — same skeleton as rng/blk: reset → ack → driver → features →
 * features_ok → setup queues → driver_ok.
 * --------------------------------------------------------------------------- */

void pci_virtio_balloon_init(void) {
  uart_println("[BALLOON] Initializing Device");

  if (!pci_find_device(VIRTIO_BALLOON_VENDOR_ID, VIRTIO_BALLOON_DEVICE_ID,
                       &bln_dev.pci)) {
    uart_println("[BALLOON] Device not found (skipping)");
    return;
  }
  uart_println("[BALLOON] Device found");

  if ((pci_get_header_type(&bln_dev.pci) & 0x7F) != PCI_ENDPOINT_DEV_TYPE) {
    uart_errorln("[BALLOON] Unexpected header type");
    return;
  }

  pci_assign_bars(&bln_dev.pci);
  pci_enable_device(&bln_dev.pci);
  virtio_parse_capabilities(&bln_dev.pci, &bln_dev.pci_caps);

  uintptr_t base = bln_dev.pci_caps.common_cfg;

  /* 1. Reset */
  mmio_write8(base + VIRTIO_COMMON_STATUS, VIRTIO_STATUS_RESET);
  dsb_sy();
  while (mmio_read8(base + VIRTIO_COMMON_STATUS) != VIRTIO_STATUS_RESET) {
  }

  /* 2. Ack + 3. Driver */
  uint8_t status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_ACKNOWLEDGE);
  dsb_sy();
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER);
  dsb_sy();

  /* 4. Features. We accept VIRTIO_F_VERSION_1 (bit 32 = feat_hi bit 0)
   * and nothing else. The balloon-specific feature bits we leave clear:
   *   - F_MUST_TELL_HOST (bit 0): we always notify before reclaiming
   *   - F_STATS_VQ (bit 1): we don't expose memory stats yet
   *   - F_DEFLATE_ON_OOM (bit 2): we'd need PMM hooks
   *   - F_FREE_PAGE_HINT / F_PAGE_POISON / F_REPORTING (bits 3..5):
   *     all advanced extensions we'd implement later. */
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 0);
  dsb_sy();
  uint32_t feat_lo = mmio_read32(base + VIRTIO_COMMON_DF);
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 1);
  dsb_sy();
  uint32_t feat_hi = mmio_read32(base + VIRTIO_COMMON_DF);
  uart_printf("[BALLOON] Device features: hi=%x lo=%x\n", feat_hi, feat_lo);

  uint32_t guest_lo = 0;
  uint32_t guest_hi = feat_hi & 0x01; /* VIRTIO_F_VERSION_1 only */

  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 0);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_lo);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 1);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_hi);
  dsb_sy();

  /* 5. FEATURES_OK + readback */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_FEATURES_OK);
  dsb_sy();
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  if (!(status & VIRTIO_STATUS_FEATURES_OK)) {
    uart_errorln("[BALLOON] FEATURES_OK rejected");
    return;
  }

  /* 6. Two virtqueues: inflate (0) + deflate (1). Same DMA setup pattern
   * for both. */
  bln_dev.inflate_vq.desc = inflate_desc;
  bln_dev.inflate_vq.avail = &inflate_avail;
  bln_dev.inflate_vq.used = &inflate_used;
  if (virtqueue_setup(base, VIRTIO_BALLOON_VQ_INFLATE, &bln_dev.inflate_vq,
                      &bln_dev.pci_caps) != ESUCCESS) {
    uart_errorln("[BALLOON] inflateq setup failed");
    return;
  }

  bln_dev.deflate_vq.desc = deflate_desc;
  bln_dev.deflate_vq.avail = &deflate_avail;
  bln_dev.deflate_vq.used = &deflate_used;
  if (virtqueue_setup(base, VIRTIO_BALLOON_VQ_DEFLATE, &bln_dev.deflate_vq,
                      &bln_dev.pci_caps) != ESUCCESS) {
    uart_errorln("[BALLOON] deflateq setup failed");
    return;
  }

  /* 7. DRIVER_OK */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER_OK);
  dsb_sy();

  bln_dev.actual = 0;
  publish_actual();
  bln_ready = 1;

  uint32_t target = mmio_read32(bln_dev.pci_caps.device_cfg +
                                VIRTIO_BALLOON_CFG_NUM_PAGES);
  uart_printf("[BALLOON] DRIVER_OK; host target=%d pages, actual=0\n",
              (uint64_t)target);
}

File: src/virtio/balloon.rs
//! VirtIO memory balloon — cooperative inflate/deflate.
//!
//! Port of `src/pci/virtio/balloon/balloon.c`. Driver-initiated (no config
//! interrupt): the shell drives inflate/deflate and /proc/balloon shows drift.

use super::virtqueue::{self, Virtqueue, VirtqAvail, VirtqDesc, VirtqUsed, VirtqUsedElem, VIRTQ_MAX_SIZE};
use super::*;
use crate::mm::mmu::virt_to_phys;
use crate::mm::pmm;
use crate::mmio;
use crate::pci;
use crate::sync::Racy;
use crate::uart;

pub const VIRTIO_BALLOON_VENDOR_ID: u16 = 0x1AF4;
pub const VIRTIO_BALLOON_DEVICE_ID: u16 = 0x1045;
const PFN_SHIFT: u64 = 12;
const MAX_PAGES: usize = 1024;
const VQ_INFLATE: u16 = 0;
const VQ_DEFLATE: u16 = 1;
const CFG_NUM_PAGES: usize = 0x00;
const CFG_ACTUAL: usize = 0x04;

#[repr(C, align(4096))]
struct BalBacking {
    inf_desc: [VirtqDesc; VIRTQ_MAX_SIZE],
    inf_avail: VirtqAvail,
    inf_used: VirtqUsed,
    def_desc: [VirtqDesc; VIRTQ_MAX_SIZE],
    def_avail: VirtqAvail,
    def_used: VirtqUsed,
    pfn_buf: [u32; MAX_PAGES],
}
const ED: VirtqDesc = VirtqDesc { addr: 0, len: 0, flags: 0, next: 0 };
const EA: VirtqAvail = VirtqAvail { flags: 0, idx: 0, ring: [0; VIRTQ_MAX_SIZE] };
const EU: VirtqUsed = VirtqUsed { flags: 0, idx: 0, ring: [VirtqUsedElem { id: 0, len: 0 }; VIRTQ_MAX_SIZE] };

static BACKING: Racy<BalBacking> = Racy::new(BalBacking {
    inf_desc: [ED; VIRTQ_MAX_SIZE], inf_avail: EA, inf_used: EU,
    def_desc: [ED; VIRTQ_MAX_SIZE], def_avail: EA, def_used: EU,
    pfn_buf: [0; MAX_PAGES],
});

struct BalDev {
    caps: VirtioPciCaps,
    inflate_vq: Virtqueue,
    deflate_vq: Virtqueue,
    actual: u32,
    inflated_pfns: [u32; MAX_PAGES],
    ready: bool,
}
static BAL: Racy<BalDev> = Racy::new(BalDev {
    caps: VirtioPciCaps { common_cfg: 0, notify_base: 0, isr_cfg: 0, device_cfg: 0, notify_off_multiplier: 0 },
    inflate_vq: Virtqueue::empty(),
    deflate_vq: Virtqueue::empty(),
    actual: 0,
    inflated_pfns: [0; MAX_PAGES],
    ready: false,
});

fn submit_pfn_batch(vq: &mut Virtqueue, count: u32) {
    if count == 0 {
        return;
    }
    let b = unsafe { BACKING.get() };
    let pa = virt_to_phys(b.pfn_buf.as_ptr() as u64);
    virtqueue::submit(vq, pa, count * 4, 0);
    virtqueue::notify(vq);
    virtqueue::poll(vq);
}

fn publish_actual() {
    let bal = unsafe { BAL.get() };
    if bal.caps.device_cfg != 0 {
        mmio::write32(bal.caps.device_cfg + CFG_ACTUAL, bal.actual);
        dsb_sy();
    }
}

pub fn inflate(mut n: u32) -> i32 {
    let bal = unsafe { BAL.get() };
    if !bal.ready {
        return -1;
    }
    let headroom = MAX_PAGES as u32 - bal.actual;
    if n > headroom {
        n = headroom;
    }
    if n == 0 {
        return 0;
    }
    let b = unsafe { BACKING.get() };
    let mut got = 0u32;
    for _ in 0..n {
        let pa = pmm::allocate_page();
        if pa == 0 {
            break;
        }
        let pfn = (pa >> PFN_SHIFT) as u32;
        b.pfn_buf[got as usize] = pfn;
        bal.inflated_pfns[(bal.actual + got) as usize] = pfn;
        got += 1;
    }
    if got == 0 {
        return 0;
    }
    submit_pfn_batch(&mut bal.inflate_vq, got);
    bal.actual += got;
    publish_actual();
    got as i32
}

pub fn deflate(mut n: u32) -> i32 {
    let bal = unsafe { BAL.get() };
    if !bal.ready {
        return -1;
    }
    if n > bal.actual {
        n = bal.actual;
    }
    if n == 0 {
        return 0;
    }
    let b = unsafe { BACKING.get() };
    let base = bal.actual - n;
    for i in 0..n {
        b.pfn_buf[i as usize] = bal.inflated_pfns[(base + i) as usize];
    }
    submit_pfn_batch(&mut bal.deflate_vq, n);
    for i in 0..n {
        pmm::free_page((b.pfn_buf[i as usize] as u64) << PFN_SHIFT);
    }
    bal.actual -= n;
    publish_actual();
    n as i32
}

/// (actual_pages, host_target).
pub fn status() -> (u32, u32) {
    let bal = unsafe { BAL.get() };
    if !bal.ready {
        return (0, 0);
    }
    let target = if bal.caps.device_cfg != 0 {
        mmio::read32(bal.caps.device_cfg + CFG_NUM_PAGES)
    } else {
        0
    };
    (bal.actual, target)
}

pub fn init() {
    uart::println("[BALLOON] Initializing Device");
    let mut pdev = match pci::find_device(VIRTIO_BALLOON_VENDOR_ID, VIRTIO_BALLOON_DEVICE_ID) {
        Some(d) => d,
        None => {
            uart::println("[BALLOON] Device not found (skipping)");
            return;
        }
    };
    pci::assign_bars(&mut pdev);
    pci::enable_device(&pdev);
    let bal = unsafe { BAL.get() };
    super::parse_capabilities(&pdev, &mut bal.caps);
    let base = bal.caps.common_cfg;
    if super::handshake(base, 0).is_none() {
        uart::errorln("[BALLOON] handshake failed");
        return;
    }
    let b = unsafe { BACKING.get() };
    bal.inflate_vq.desc = b.inf_desc.as_mut_ptr();
    bal.inflate_vq.avail = &mut b.inf_avail as *mut VirtqAvail;
    bal.inflate_vq.used = &mut b.inf_used as *mut VirtqUsed;
    if !virtqueue::setup(base, VQ_INFLATE, &mut bal.inflate_vq, &bal.caps) {
        uart::errorln("[BALLOON] inflate queue setup failed");
        return;
    }
    bal.deflate_vq.desc = b.def_desc.as_mut_ptr();
    bal.deflate_vq.avail = &mut b.def_avail as *mut VirtqAvail;
    bal.deflate_vq.used = &mut b.def_used as *mut VirtqUsed;
    if !virtqueue::setup(base, VQ_DEFLATE, &mut bal.deflate_vq, &bal.caps) {
        uart::errorln("[BALLOON] deflate queue setup failed");
        return;
    }
    super::set_driver_ok(base);
    bal.ready = true;
    uart::println("[BALLOON] DRIVER_OK");
}
